File: src/03.number/32.372.go
package number_bit

var base int = 1337

func myPow(a, k int) int {
	if k == 0 {
		return 1
	}
	a %= base

	if k%2 == 1 {
		return (a * myPow(a, k-1)) % base
	} else {
		sub := myPow(a, k/2)
		return (sub * sub) % base
	}
}

func SuperPow(a int, b []int) int {
	if len(b) <= 0 {
		return 1
	}

	last := b[len(b)-1]
	ans1 := myPow(a, last)
	ans2 := myPow(SuperPow(a, b[:len(b)-1]), 10)

	return (ans1 * ans2) % base
}

File: src/03.number/32.372.ts
const myPow = (m: number, n: number): number => {
    // let res = 1;
    // m %= 1337;
    // for (let i = 0; i < n; i++) {
    //     res *= m;
    //     res %= 1337;
    // }
    // return res;

    // 高效求幂
    // a^b = ?
    // b 为奇数时, a^b = a * a^(b-1)
    // b 为偶数时, (a^(b/2))^2
    if (n === 0) {
        return 1;
    }

    m %= 1337;
    if (n % 2 === 1) {
        return (m * myPow(m, n - 1)) % 1337;
    } else {
        const sub = myPow(m, n / 2);
        return (sub * sub) % 1337;
    }
};

export function superPow(a: number, b: number[]): number {
    // a^112 = a^2 * a^110 = a^2 * (a^11)^10
    // a*b*c%d = (a%d * b%d * c%d) %d
    // 对乘法的结果求模，等价于先对每个因子都求模，然后对因子相乘的结果再求模。

    const bLen = b.length;

    if (!bLen) {
        return 1;
    }

    const last = b.pop();
    const ans1 = myPow(a, last);
    const ans2 = myPow(superPow(a, b), 10);

    return (ans1 * ans2) % 1337;
}
